File: src/day07.rs
use std::{fs, iter::zip, num::ParseIntError, path::PathBuf, str::FromStr};

use itertools::Itertools;

pub fn day07(mut input_path: PathBuf) {
    input_path.push("07.txt");
    let content = fs::read_to_string(input_path).unwrap();

    let equations: Vec<Equation> = content.lines().map(|l| l.parse().unwrap()).collect();

    println!("{:?}", part1(&equations));
    println!("{:?}", part2(&equations));
}

#[derive(Debug)]
struct Equation {
    result: u64,
    terms: Vec<u64>,
}

#[derive(Debug, Clone, Copy)]
enum Operator {
    Add,
    Multiply,
    Concat,
}

fn part1(equations: &[Equation]) -> u64 {
    equations
        .iter()
        .filter(|e| can_be_true(e, vec![Operator::Add, Operator::Multiply]))
        .map(|e| e.result)
        .sum()
}

fn part2(equations: &[Equation]) -> u64 {
    equations
        .iter()
        .filter(|e| can_be_true(e, vec![Operator::Add, Operator::Multiply, Operator::Concat]))
        .map(|e| e.result)
        .sum()
}

fn eval(a: u64, op: Operator, b: u64) -> u64 {
    match op {
        Operator::Add => a + b,
        Operator::Multiply => a * b,
        Operator::Concat => {
            let mut o = 1;
            while o <= b {
                o *= 10;
            }
            a * o + b
        }
    }
}

fn is_true(equation: &Equation, operators: &[Operator]) -> bool {
    let [init, tail @ ..] = equation.terms.as_slice() else {
        // we checked during parsing
        panic!("RHS needs at least one term");
    };

    equation.result == zip(operators, tail).fold(*init, |acc, (&op, &t)| eval(acc, op, t))

    // Early stopping the fold when we exceed result is surprisingly not appreciably faster.
    // this try_fold is similar to (sum . filter ((>) result) . (scan foldop))

    // zip(operators, tail)
    //     .try_fold(*init, |acc, (&op, &t)| {
    //         let s = eval(acc, op, t);
    //         if s > equation.result {
    //             return Err(s);
    //         }
    //         Ok(s)
    //     })
    //     .is_ok_and(|lhs| lhs == equation.result)
}

fn can_be_true(equation: &Equation, options: Vec<Operator>) -> bool {
    operator_combinations_for(&equation, options).any(|ops| is_true(&equation, &ops))
}

fn operator_combinations_for(
    equation: &Equation,
    options: Vec<Operator>,
) -> Combinations<Operator> {
    Combinations::new((equation.terms.len() - 1) as u32, options)
}

struct Combinations<T> {
    // TODO: Don't need to own that
    options: Vec<T>,
    length: u32,
    current: usize,
}

impl<T> Combinations<T> {
    fn new(length: u32, options: Vec<T>) -> Self {
        Combinations {
            options,
            length,
            current: 0,
        }
    }
}

impl<T: Copy> Iterator for Combinations<T> {
    type Item = Vec<T>;

    fn next(&mut self) -> Option<Self::Item> {
        let k = self.options.len();
        self.current += 1;
        if self.current > k.pow(self.length) {
            return None;
        }
        let mut v: Vec<T> = Vec::with_capacity(self.length as usize);
        let n = self.current - 1;
        for i in 0..self.length {
            // part 1 was easy, we could map bits to operation: n >> i & 1
            v.push(self.options[(n / k.pow(i)) % k]);
        }
        Some(v)
    }
}

#[derive(Debug)]
struct ParseEquationError;

// so we can use '?' on Result<_, ParseIntError> in a function returning Result<_, ParseEquationError>
impl From<ParseIntError> for ParseEquationError {
    fn from(_: ParseIntError) -> Self {
        ParseEquationError
    }
}

impl FromStr for Equation {
    type Err = ParseEquationError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let Some((lhs, rhs)) = s.split(":").next_tuple() else {
            return Err(ParseEquationError);
        };
        let result = lhs.parse()?;
        let terms = rhs
            .split_whitespace()
            .map(|t| t.parse::<u64>())
            .collect::<Result<Vec<_>, _>>()?;
        match terms.len() {
            0 => Err(ParseEquationError),
            _ => Ok(Equation { result, terms }),
        }
    }
}

File: src/day07.hs
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE ViewPatterns #-}

import Control.Monad.Zip (mzip)
import Data.List (singleton)
import Data.List.Split (splitOn)
import Data.Tree

data Equation = Equation {result :: Int, terms :: [Int]} deriving (Show)

instance Read Equation where
  readsPrec _ (splitOn ":" -> [lhs, rhs]) =
    let result = read lhs
        terms = map read $ words rhs
     in [(Equation {result = result, terms = terms}, "")]
  readsPrec _ _ = []

data Operator = Add | Multiply | Concat | Init deriving (Show)

operators :: Tree Operator
operators = unfoldTree (,[Add, Multiply, Concat]) Init

toLayers :: [a] -> Tree a
toLayers (x : xs) = unfoldTree f (x, xs)
  where
    f (current, []) = (current, [])
    f (current, next : rest) = (current, repeat (next, rest))

evalExpr :: Int -> Operator -> Int -> Int
evalExpr _ Init a = a
evalExpr a Add b = a + b
evalExpr a Multiply b = a * b
evalExpr a Concat b = a * o + b
  where
    o = head (dropWhile (b >=) [10 ^ x | x <- [0 ..]])

interimResults :: [Int] -> Tree Int
interimResults terms = scanTree (uncurry . evalExpr) 0 (mzip operators (toLayers terms))

isSatisfyable :: Equation -> Bool
isSatisfyable (Equation r ts) = elem r $ takeWhileTree (r >=) $ interimResults ts

main = do
  contents <- readFile "input7.txt"
  let equations = (map read $ lines contents) :: [Equation]
  putStrLn "Part 2:"
  print $ sum $ map result $ filter isSatisfyable equations

scanTree :: (b -> a -> b) -> b -> Tree a -> Tree b
scanTree f s t = unfoldTree unfoldF (s, t)
  where
    unfoldF (acc, Node r sf) = (res, map (res,) sf)
      where
        res = f acc r

takeWhileTree :: (a -> Bool) -> Tree a -> Tree a
takeWhileTree p = head . unfoldForest unfoldF . singleton
  where
    unfoldF (Node r sf) = (r, filter (p . rootLabel) sf)
